# Data Preprocessing: Merging, Cleaning, and Feature Engineering for Movie Dataset

Ithis fileon, we perform several important data preprocessing steps to prepare the datasefor visualizationsns and Modetrainingis. This involves loading anmerging two differentle datasets, handling missing values, and engineering nefeatures.`.

In [15]:
import numpy as np
import pandas as pd
import pandas as pd
import numpy as np

### 1. **Loading and Merging Datasets**
The code starts by importing the necessary libraries, `pandas` and `numpy`, to work with data. Two CSV files, `tmdb_5000_credits.csv` and `tmdb_5000_movies.csv`, are loaded into two separate DataFrames: `credits` and `movies`.

In [17]:
# Load the datasets
credits = pd.read_csv("tmdb_5000_credits.csv")
movies = pd.read_csv("tmdb_5000_movies.csv")

# Merge the datasets on the 'title' column
df = movies.merge(credits, on="title")

# Save the merged dataset to a new CSV file
df.to_csv("MoviesData.csv", index=False)

print("Merged file saved as MoviesData.csv")

Merged file saved as MoviesData.csv


Merged file is saved in MoviesData.csv to perform preprocessing and feature engineering steps. Along with the saving dataframe 'df' is created for further use.

In [19]:
# Load the dataset
import pandas as pd

# Load the dataset
df = pd.read_csv('MoviesData.csv')

# Show the first few rows to get an overview of the dataset
df.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

The below rows perform exploratory analysis on the 'df' dataframe. This section helps in understanding the structure of data.

In [21]:
df.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'movie_id', 'cast', 'crew'],
      dtype='object')

In [22]:
# Get a summary of the movies dataset
print(df.info())

# Get descriptive statistics for numeric features
print(movies.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [23]:
df=df.loc[df['overview'].notnull()] 
df=df.loc[df['release_date'].notnull()]

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4805 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4805 non-null   int64  
 1   genres                4805 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4805 non-null   int64  
 4   keywords              4805 non-null   object 
 5   original_language     4805 non-null   object 
 6   original_title        4805 non-null   object 
 7   overview              4805 non-null   object 
 8   popularity            4805 non-null   float64
 9   production_companies  4805 non-null   object 
 10  production_countries  4805 non-null   object 
 11  release_date          4805 non-null   object 
 12  revenue               4805 non-null   int64  
 13  runtime               4805 non-null   float64
 14  spoken_languages      4805 non-null   object 
 15  status                4805

Handling missing values is crucial for unbiased analysis and model training. Here we first understand the number of missing values in the dataset.

In [26]:
print(df.isnull().sum())

budget                     0
genres                     0
homepage                3092
id                         0
keywords                   0
original_language          0
original_title             0
overview                   0
popularity                 0
production_companies       0
production_countries       0
release_date               0
revenue                    0
runtime                    0
spoken_languages           0
status                     0
tagline                  840
title                      0
vote_average               0
vote_count                 0
movie_id                   0
cast                       0
crew                       0
dtype: int64


According to the output, we can see that the columns 'homepage' and 'tagline' have 3092 and 840 missing values. In this project, the column 'homepage' will not be used for recommendations; hence, we will drop the column. The column 'tagline' is important for the content-based filtering; therefore, we will impute missing values in the 'tagline' column with the 'title' of that movie

In [28]:
df['tagline'] = df['tagline'].fillna(df['title'])
df.drop('homepage', axis=1, inplace=True)

In [29]:
print(df.isnull().sum())

budget                  0
genres                  0
id                      0
keywords                0
original_language       0
original_title          0
overview                0
popularity              0
production_companies    0
production_countries    0
release_date            0
revenue                 0
runtime                 0
spoken_languages        0
status                  0
tagline                 0
title                   0
vote_average            0
vote_count              0
movie_id                0
cast                    0
crew                    0
dtype: int64


In [30]:
df.head()

budget                                             genres      id  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries  ... runtime  \
0  [{"iso_3166_1": "US", "name": "United States o...  ...   162.0   
1  [{"iso_3166_1": "US", "name": "United States o...  ...   169.0   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...  ...   148.0   
3  [{"iso_3166_1": "US", "name": "United States o...  ...   165.0   
4  [{"iso_3166_1": "US", "name": "United States o...  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count  movie_id  \
0                                    Avatar          7.2      11800     19995   
1  Pirates of the Caribbean: At World's End          6.9       4500       285   
2                                   Spectre          6.3       4466    206647   
3                     The Dark Knight Rises          7.6       9106     49026   
4                               John Carter          6.1       2124     49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id"

For further feature engineering, we will convert the 'release_date' column to datetime type and extract year from the column. This will be helpful while performing collaborative filtering recommendations.

In [32]:
df['release_date']=pd.to_datetime(df['release_date'])

In [33]:
df['release_date']

0      2009-12-10
1      2007-05-19
2      2015-10-26
3      2012-07-16
4      2012-03-07
          ...    
4804   1992-09-04
4805   2011-12-26
4806   2013-10-13
4807   2012-05-03
4808   2005-08-05
Name: release_date, Length: 4805, dtype: datetime64[ns]

In [34]:
df['year'] = df['release_date'].dt.year

In [35]:
df['year']

0       2009
1       2007
2       2015
3       2012
4       2012
        ... 
4804    1992
4805    2011
4806    2013
4807    2012
4808    2005
Name: year, Length: 4805, dtype: int32

In this code, we aim to rank movies based on a **weighted rating system** that balances both popularity (number of votes) and quality (average rating). The first step is to calculate **C**, the overall mean vote average of all movies in the dataset. This serves as a baseline for adjusting ratings. Next, we determine **m**, the minimum number of votes required for a movie to be considered. This threshold is set at the **90th percentile** of vote counts, ensuring that only the most popular movies are ranked. Movies with fewer votes than **m** are filtered out to prevent unfair ranking due to limited reviews.
In below formula
where **v** is the number of votes a movie has received, **R** is its average rating, **m** is the minimum votes required, and **C** is the mean rating across all movies. This formula ensures that movies with a high number of votes get more influence, while those with fewer votes are adjusted toward the mean. The function is applied to the filtered dataset, creating a new **weighted_rating** clumn.


In [61]:
# Calculate overall mean vote average (C)
C = df['vote_average'].mean()

# Define m as the 90th percentile of vote_count
m = df['vote_count'].quantile(0.90)

# Function to compute the weighted rating
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m)) * R + (m/(v+m)) * C

# Apply weighted rating to all movies, filling NaN values for those with low vote_count
df['weighted_rating'] = df.apply(lambda x: weighted_rating(x) if x['vote_count'] >= m else np.nan, axis=1)

# Sorting movies by weighted rating (excluding NaN values)
demographic_recommendations = df.dropna(subset=['weighted_rating']).sort_values('weighted_rating', ascending=False)

# Display top 10 movies based on weighted rating
print(demographic_recommendations[['original_title', 'vote_average', 'vote_count', 'weighted_rating']].head(10))

                                     original_title  vote_average  vote_count  \
1887                       The Shawshank Redemption           8.5        8205   
662                                      Fight Club           8.3        9413   
65                                  The Dark Knight           8.2       12002   
3237                                   Pulp Fiction           8.3        8428   
96                                        Inception           8.1       13752   
3342                                  The Godfather           8.4        5893   
95                                     Interstellar           8.1       10867   
809                                    Forrest Gump           8.2        7927   
329   The Lord of the Rings: The Return of the King           8.1        8064   
1996                        The Empire Strikes Back           8.2        5879   

      weighted_rating  
1887         8.058086  
662          7.938297  
65           7.919252  
3237        

In [63]:
df

budget                                             genres      id  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   
...         ...                                                ...     ...   
4804     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...    9367   
4805       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   72766   
4806          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...  231617   
4807          0                                                 []  126186   
4808          0                [{"id": 99, "name": "Documentary"}]   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4804  [{"id": 5616, "name": "united states\u2013mexi...                es   
4805                                                 []                en   
4806  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4807                                                 []                en   
4808  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4804                               El Mariachi   
4805                                 Newlyweds   
4806                 Signed, Sealed, Delivered   
4807                          Shanghai Calling   
4808                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4804  El Mariachi just wants to play his guitar and ...   14.269792   
4805  A newlywed couple's honeymoon is upended by th...    0.642552   
4806  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4807  When ambitious New York attorney Sam is sent t...    0.857008   
4808  Ever since the second grade when he first saw ...    1.929883   

                                   production_companies  \
0     [{"name": "Ingenious Film Partners", "id": 289...   
1     [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2     [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3     [{"name": "Legendary Pictures", "id": 923}, {"...   
4           [{"name": "Walt Disney Pictures", "id": 2}]   
...                                                 ...   
4804           [{"name": "Columbia Pictures", "id": 5}]   
4805                                                 []   
4806  [{"name": "Front Street Pictures", "id": 3958}...   
4807                                                 []   
4808  [{"name": "rusty bear entertainment", 

Finally we will save the Processed and Engineered data to **MoviesData_Processed.csv** for further model training

In [65]:
# Save the processed and engineered dataset to a new CSV file
df.to_csv("MoviesData_Processed.csv", index=False)